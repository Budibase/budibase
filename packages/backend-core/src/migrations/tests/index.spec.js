require("../../../tests")
const { runMigrations, getMigrationsDoc } = require("../index")
const { getDB } = require("../../db")
const {
  StaticDatabases,
} = require("../../constants")

let db

describe("migrations", () => {

  const migrationFunction = jest.fn()

  const MIGRATIONS = [{
    type: "global",
    name: "test",
    fn: migrationFunction
  }]

  beforeEach(() => {
    db = getDB(StaticDatabases.GLOBAL.name)
  })

  afterEach(async () => {
    jest.clearAllMocks()
    await db.destroy()
  })

  const migrate = () => {
    return runMigrations(MIGRATIONS)
  }

  it("should run a new migration", async () => {
    await migrate()
    expect(migrationFunction).toHaveBeenCalled()
    const doc = await getMigrationsDoc(db)
    expect(doc.test).toBeDefined()
  })

  it("should match snapshot", async () => {
    await migrate()
    const doc = await getMigrationsDoc(db)
    expect(doc).toMatchSnapshot()
  })

  it("should skip a previously run migration", async () => {
    await migrate()
    const previousMigrationTime = await getMigrationsDoc(db).test
    await migrate()
    const currentMigrationTime = await getMigrationsDoc(db).test
    expect(migrationFunction).toHaveBeenCalledTimes(1)
    expect(currentMigrationTime).toBe(previousMigrationTime)
  })
})