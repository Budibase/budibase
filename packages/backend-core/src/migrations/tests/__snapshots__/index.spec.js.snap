// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrations should match snapshot 1`] = `
Object {
  "_id": "migrations",
  "_rev": "1-a32b0b708e59eeb006ed5e063cfeb36a",
  "createdAt": "2020-01-01T00:00:00.000Z",
  "test": 1577836800000,
  "updatedAt": "2020-01-01T00:00:00.000Z",
}
`;
